src/mpid/ch4/src/mpidig_part_callbacks.c

```diff
@@ -141,27 +170,48 @@ int MPIDIG_part_send_data_target_msg_cb(void *am_hdr, void *data,
     rreq->comm = part_rreq->comm;
     MPIR_Comm_add_ref(rreq->comm);
 
-    MPIDIG_REQUEST(rreq, buffer) = MPIDI_PART_REQUEST(part_rreq, buffer);
+    //  get the right partition location
+    const int ipart = msg_hdr->ipart;
+    MPIR_Assert(ipart >= 0);
+    MPIR_Assert(ipart < MPIDIG_PART_REQUEST(part_rreq, u.recv.send_part));
+
+    // get the count from the present request
+    int count = MPIDI_PART_REQUEST(part_rreq, count);
+
+    // get the exact buffer location given the ipart and the offset
+    MPI_Aint part_offset;
+    MPIR_Datatype_get_size_macro(MPIDI_PART_REQUEST(part_rreq, datatype), part_offset);
```

Comment:
I think you want the extent here instead of the size? Size only counts the bytes used, not gaps/padding.

Reply:
oh that's right, it's sneaky

Reply:
solved in 87ccea181bd6cf17abb739abcd610e17782195f4